// @observablehq/framework/dist/client/stdlib/recommendedLibraries.js
const _ = () => import("npm:lodash").then((lodash) => lodash.default);
const aq = () => import("npm:arquero");
const Arrow = () => import("npm:apache-arrow");
const d3 = () => import("npm:d3");
const dot = () => import("observablehq:stdlib/dot").then((dot2) => dot2.default);
const duckdb = () => import("npm:@duckdb/duckdb-wasm");
const DuckDBClient = () => import("observablehq:stdlib/duckdb").then((duckdb2) => duckdb2.DuckDBClient);
const echarts = () => import("npm:echarts");
const htl = () => import("npm:htl");
const html = () => import("npm:htl").then((htl2) => htl2.html);
const svg = () => import("npm:htl").then((htl2) => htl2.svg);
const Inputs = () => import("npm:@observablehq/inputs");
const L = () => import("npm:leaflet");
const mapboxgl = () => import("npm:mapbox-gl").then((module) => module.default);
const mermaid = () => import("observablehq:stdlib/mermaid").then((mermaid2) => mermaid2.default);
const Plot = () => import("npm:@observablehq/plot");
const React = () => import("npm:react");
const ReactDOM = () => import("npm:react-dom");
const sql = () => import("observablehq:stdlib/duckdb").then((duckdb2) => duckdb2.sql);
const SQLite = () => import("observablehq:stdlib/sqlite").then((sqlite) => sqlite.default);
const SQLiteDatabaseClient = () => import("observablehq:stdlib/sqlite").then((sqlite) => sqlite.SQLiteDatabaseClient);
const tex = () => import("observablehq:stdlib/tex").then((tex2) => tex2.default);
const topojson = () => import("npm:topojson-client");
const vg = () => import("observablehq:stdlib/vgplot").then((vg2) => vg2.default());
const vl = () => import("observablehq:stdlib/vega-lite").then((vl2) => vl2.default);
export {
  Arrow,
  DuckDBClient,
  Inputs,
  L,
  Plot,
  React,
  ReactDOM,
  SQLite,
  SQLiteDatabaseClient,
  _,
  aq,
  d3,
  dot,
  duckdb,
  echarts,
  htl,
  html,
  mapboxgl,
  mermaid,
  sql,
  svg,
  tex,
  topojson,
  vg,
  vl
};
